File: AndroidDatabase/DataWithAdapter/app/src/main/java/com/haoyudu/roomdatabaseupgrade/WordDataBase.java
package com.haoyudu.roomdatabaseupgrade;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

@Database(entities = Word.class, version = 4, exportSchema = false)
public abstract class WordDataBase extends RoomDatabase {
    private static WordDataBase instance;

    static synchronized WordDataBase getInstance(Context context) {
        if (instance == null) {
            instance = Room.databaseBuilder(context.getApplicationContext(), WordDataBase.class, "word_database").build();
        }
        return instance;
    }

    public abstract WordDao getWordDao();

}

File: KotlinTest/KotlinDatabase/app/src/main/java/com/haoyudu/kotlindatabase/WordDatabase.kt
package com.haoyudu.kotlindatabase

import android.app.Application
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase

@Database(entities = [Word::class] ,version=1,exportSchema = false)
abstract class WordDatabase: RoomDatabase() {
    abstract fun getDao():WordDao

    companion object{
        private  var database: WordDatabase? =null
        @Synchronized
        @JvmStatic
        fun getInstance(application:Application):WordDatabase{
            if (database==null){
                database= Room.databaseBuilder(application,WordDatabase::class.java,"word_database").build()
            }
            return database as WordDatabase
        }
    }

}